module/zfs/dsl_crypt.c

```diff
@@ -2696,8 +2697,15 @@ spa_do_crypt_objset_mac_abd(boolean_t generate, spa_t *spa, uint64_t dsobj,
 		return (0);
 	}
 
+	intval = osp->os_flags;
+	if (byteswap)
+		intval = BSWAP_64(intval);
+
+	/* Do not compare the local_mac if user accounting is not complete */
 	if (bcmp(portable_mac, osp->os_portable_mac, ZIO_OBJSET_MAC_LEN) != 0 ||
-	    bcmp(local_mac, osp->os_local_mac, ZIO_OBJSET_MAC_LEN) != 0) {
+	    ((intval & OBJSET_FLAG_USERACCOUNTING_COMPLETE) != 0 &&
+	    (intval & OBJSET_FLAG_USEROBJACCOUNTING_COMPLETE) != 0 &&
+	    bcmp(local_mac, osp->os_local_mac, ZIO_OBJSET_MAC_LEN) != 0)) {
```

Comment:
I think it would be worthwhile to break the local_mac check in to its own conditional block for readability. It seems like we should also be checking to make sure the feature is enabled and that it's complete.  i.e. `dmu_objset_userused_enabled()` and `dmu_objset_userobjused_enabled()`.